tests/Unit/Helpers/Evolution/Systems/GrMhd/ValenciaDivClean/FiniteDifference/PrimReconstructor.hpp

```diff
@@ -0,0 +1,355 @@
+    tmpl::for_each<tmpl::list<Rho, Pressure, MagField, Phi>>(
+        [&volume_prims, &volume_prims_for_recons](auto tag_v) {
+          using tag = tmpl::type_from<decltype(tag_v)>;
+          get<tag>(volume_prims) = get<tag>(volume_prims_for_recons);
+        });
+    // Metric is identity at cell center
+    get(get<LorentzFactor>(volume_prims)) =
+        sqrt(1.0 + get(dot_product(get<VelocityW>(volume_prims_for_recons),
+                                   get<VelocityW>(volume_prims_for_recons))));
+    for (size_t i = 0; i < 3; ++i) {
+      get<Velocity>(volume_prims).get(i) =
+          get<VelocityW>(volume_prims_for_recons).get(i) /
+          get(get<LorentzFactor>(volume_prims));
+    }
+  }
+
+  // Now we have everything to call the reconstruction
+  dynamic_cast<const Reconstructor&>(reconstructor)
+      .reconstruct(make_not_null(&vars_on_lower_face),
+                   make_not_null(&vars_on_upper_face), volume_prims, eos,
+                   element, neighbor_data, subcell_mesh);
+
+  for (size_t dim = 0; dim < 3; ++dim) {
+    CAPTURE(dim);
+    const auto basis = make_array<3>(Spectral::Basis::FiniteDifference);
+    auto quadrature = make_array<3>(Spectral::Quadrature::CellCentered);
+    auto extents = make_array<3>(points_per_dimension);
+    gsl::at(extents, dim) = points_per_dimension + 1;
+    gsl::at(quadrature, dim) = Spectral::Quadrature::FaceCentered;
+    const Mesh<3> face_centered_mesh{extents, basis, quadrature};
+    auto logical_coords_face_centered = logical_coordinates(face_centered_mesh);
+    for (size_t i = 1; i < 3; ++i) {
+      logical_coords_face_centered.get(i) =
+          logical_coords_face_centered.get(i) + 4.0 * i;
+    }
+    Variables<dg_package_data_argument_tags> expected_lower_face_values{
+        face_centered_mesh.number_of_grid_points()};
+    expected_lower_face_values.assign_subset(
+        compute_solution(logical_coords_face_centered));
+    if constexpr (ThermodynamicDim == 2) {
+      get<SpecificInternalEnergy>(expected_lower_face_values) =
+          eos.specific_internal_energy_from_density_and_pressure(
+              get<Rho>(expected_lower_face_values),
+              get<Pressure>(expected_lower_face_values));
+    } else {
+      get<SpecificInternalEnergy>(expected_lower_face_values) =
+          eos.specific_internal_energy_from_density(
+              get<Rho>(expected_lower_face_values));
+    }
+    get<SpecificEnthalpy>(expected_lower_face_values) =
+        hydro::relativistic_specific_enthalpy(
+            get<Rho>(expected_lower_face_values),
+            get<SpecificInternalEnergy>(expected_lower_face_values),
+            get<Pressure>(expected_lower_face_values));
+    get<gr::Tags::SqrtDetSpatialMetric<DataVector>>(
+        expected_lower_face_values) = lower_face_sqrt_det_spatial_metric;
+    get<gr::Tags::SpatialMetric<3>>(expected_lower_face_values) =
+        lower_face_spatial_metric;
```

Comment:
[nitpick] Keeping track of the lower vs upper would be easier if lines 266-269 were adjacent to lines 289-292... is there a reason this first block can't be moved down above line 289?